src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+            if (_serverCertificateValidationCallback != null)
+            {
+                optionData =
+                    Interop.WinHttp.SECURITY_FLAG_IGNORE_UNKNOWN_CA |
+                    Interop.WinHttp.SECURITY_FLAG_IGNORE_CERT_WRONG_USAGE |
+                    Interop.WinHttp.SECURITY_FLAG_IGNORE_CERT_CN_INVALID |
+                    Interop.WinHttp.SECURITY_FLAG_IGNORE_CERT_DATE_INVALID;
+                SetWinHttpOption(requestHandle, Interop.WinHttp.WINHTTP_OPTION_SECURITY_FLAGS, ref optionData);
+            }
+            else if (_checkCertificateRevocationList)
+            {
+                // If no custom validation method, then we let WinHTTP do the revocation check itself.
+                optionData = Interop.WinHttp.WINHTTP_ENABLE_SSL_REVOCATION;
+                SetWinHttpOption(requestHandle, Interop.WinHttp.WINHTTP_OPTION_ENABLE_FEATURE, ref optionData);
+            }
+        }
+
+        private void SetRequestHandleClientCertificateOptions(SafeInternetHandle requestHandle, Uri requestUri)
+        {
+            // Must be HTTPS scheme to use client certificates.
+            if (requestUri.Scheme != UriSchemeHttps)
+            {
+                return;
+            }
+
+            // Get candidate list for client certificates.
+            X509Certificate2Collection certs;
+            if (_clientCertificateOption == ClientCertificateOption.Manual)
+            {
+                certs = ClientCertificates;
+            }
+            else
+            {
+                var myStore = new X509Store();
+                myStore.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);
+                certs = myStore.Certificates;
+                myStore.Dispose();
+            }
+
+            // Check for no certs now as a performance optimization.
+            if (certs.Count == 0)
+            {
+                SetNoClientCertificate(requestHandle);
+                return;
+            }
+
+            // Reduce the set of certificates to match the proper 'Client Authentication' criteria.
+            certs = certs.Find(X509FindType.FindByKeyUsage, X509KeyUsageFlags.DigitalSignature, true);
+            certs = certs.Find(X509FindType.FindByApplicationPolicy, ClientAuthenticationOID, true);
+
+            // Build a new collection with certs that have a private key. Need to do this
+            // manually because there is no X509FindType to match this criteria.
+            var clientCerts = new X509Certificate2Collection();
+            foreach (var cert in certs)
+            {
+                if (cert.HasPrivateKey)
+                {
+                    clientCerts.Add(cert);
+                }
+            }
+
+            // TOOD: Filter the list based on TrustedIssuerList info from WINHTTP.
+
+            // Set the client certificate.
+            if (certs.Count == 0)
+            {
+                SetNoClientCertificate(requestHandle);
+            }
+            el
```

Comment:
@FiveTimesTheFun / @KrzysztofCwalina, we should spend some time thinking about #1991.  It'd be great if we could do something to avoid the large allocations for buffers we have in these streams, so as to help in cases like this where it's being used per request and pooling-based approach isn't really feasible.